File: backend/Models/Rounds/Round.cs
using Models.Specifications;

namespace Models.Rounds;

public class Round : EntityInfo, IEntity
{
    public Settings.Settings Settings { get; set; } = null!;
    public Guid? WinnerId => Participants.FirstOrDefault(x => x.IsWinner)?.TeamId;
    public Specification? Specification { get; set; }
    public List<Participant> Participants { get; set; } = new();
    public List<Artifact> Artifacts { get; set; } = new();
    public List<HistoryItem> History { get; set; } = new();
    public HistoryItem? CurrentState => History.MaxBy(x => x.Order);
    public Guid? NextRoundId { get; set; }
    public int Order { get; set; }
    public Guid GameId { get; set; }
    public string? Description { get; set; }
}

File: frontend/src/types/round.ts
import { TourneySettings, TourneySpecification } from "~/types/tourney";

export type Round = {
  id: string;
  description?: string;
  nextRoundId?: string;
  gameId: string;
  settings: TourneySettings;
  specification: TourneySpecification;
  participants: RoundParticipant[];
  artifacts: unknown[];
  history: RoundStateData[];
  winnerId?: string;
  currentState?: RoundStateData;
  order: number;
};

export type CreateRound = {
  gameId: string;
  nextRoundId?: string | null;
  participants: RoundParticipant[];
  specification: TourneySpecification;
  settings: TourneySettings;
  order: number;
};

export type FinishRound = {
  marks: RoundMark[];
};

export type RoundStateData = {
  state?: RoundState;
  value?: {
    currentState?: RoundState;
    start?: Date;
    teamId?: string;
  };
  order: number;
};

export enum RoundState {
  Prepare = "prepare",
  Presentation = "presentation",
  Defense = "defense",
  Mark = "mark",
  Complete = "complete",
  Pause = "pause",
}

export type RoundParticipant = {
  teamId: string;
  challenges: string[];
  isWinner: boolean;
  points: number;
  order: number;
};

export type RoundMark = {
  teamId: string;
  mark: number;
};
